File: android/app/src/main/java/org/fossasia/openevent/utils/JWTUtils.java
package org.fossasia.openevent.utils;

import android.support.v4.util.SparseArrayCompat;

import org.json.JSONException;
import org.json.JSONObject;

public class JWTUtils {

    public static SparseArrayCompat<String> decode(String token) {
        SparseArrayCompat<String> decoded = new SparseArrayCompat<>(2);

        String[] split = token.split("\\.");
        decoded.append(0, getJson(split[0]));
        decoded.append(1, getJson(split[1]));

        return decoded;
    }

    public static long getExpiry(String token) throws JSONException {
        SparseArrayCompat<String> decoded = decode(token);

        // We are using JSONObject instead of GSON as it takes about 5 ms instead of 150 ms taken by GSON
        return Long.parseLong(new JSONObject(decoded.get(1)).get("exp").toString());
    }

    public static boolean isExpired(String token) {
        long expiry;

        try {
            expiry = getExpiry(token);
        } catch (JSONException jse) {
            return true;
        }

        return System.currentTimeMillis() / 1000 >= expiry;
    }

    private static String getJson(String strEncoded) {
        byte[] decodedBytes = Base64Utils.decode(strEncoded);
        return new String(decodedBytes);
    }

    /**
     * Base64 class because we can't test Android class and this is faster
     */
    private static class Base64Utils {

        private final static char[] ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/".toCharArray();

        private static int[] toInt = new int[128];

        static {
            for (int i = 0; i < ALPHABET.length; i++) {
                toInt[ALPHABET[i]] = i;
            }
        }

        /**
         * Translates the specified Base64 string into a byte array.
         *
         * @param s the Base64 string (not null)
         * @return the byte array (not null)
         */
        private static byte[] decode(String s) {
            int delta = s.endsWith("==") ? 2 : s.endsWith("=") ? 1 : 0;
            byte[] buffer = new byte[s.length() * 3 / 4 - delta];
            int mask = 0xFF;
            int index = 0;
            for (int i = 0; i < s.length(); i += 4) {
                int c0 = toInt[s.charAt(i)];
                int c1 = toInt[s.charAt(i + 1)];
                buffer[index++] = (byte) (((c0 << 2) | (c1 >> 4)) & mask);
                if (index >= buffer.length) {
                    return buffer;
                }
                int c2 = toInt[s.charAt(i + 2)];
                buffer[index++] = (byte) (((c1 << 4) | (c2 >> 2)) & mask);
                if (index >= buffer.length) {
                    return buffer;
                }
                int c3 = toInt[s.charAt(i + 3)];
                buffer[index++] = (byte) (((c2 << 6) | c3) & mask);
            }
            return buffer;
        }

    }
}

File: android/app/src/main/java/org/fossasia/openevent/common/api/JWTUtils.kt
package org.fossasia.openevent.common.api

import android.support.v4.util.SparseArrayCompat

import org.json.JSONException
import org.json.JSONObject

object JWTUtils {

    private fun decode(token: String): SparseArrayCompat<String> {
        val decoded = SparseArrayCompat<String>(2)

        val split = token.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        decoded.append(0, getJson(split[0]))
        decoded.append(1, getJson(split[1]))

        return decoded
    }

    @Throws(JSONException::class)
    private fun getExpiry(token: String): Long {
        val decoded = decode(token)

        // We are using JSONObject instead of GSON as it takes about 5 ms instead of 150 ms taken by GSON
        return JSONObject(decoded.get(1)).get("exp").toString().toLong()
    }

    @Throws(JSONException::class)
    @JvmStatic
    fun getIdentity(token: String): Int {
        val decoded = decode(token)

        return JSONObject(decoded.get(1)).get("identity").toString().toInt()
    }

    @JvmStatic
    fun isExpired(token: String): Boolean {
        val expiry: Long

        try {
            expiry = getExpiry(token)
        } catch (jse: JSONException) {
            return true
        }

        return System.currentTimeMillis() / 1000 >= expiry
    }

    private fun getJson(strEncoded: String): String {
        val decodedBytes = Base64Utils.decode(strEncoded)
        return String(decodedBytes)
    }

    /**
     * Base64 class because we can't test Android class and this is faster
     */
    private object Base64Utils {

        private val ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/".toCharArray()

        private val toInt = IntArray(128)

        init {
            for (i in ALPHABET.indices) {
                toInt[ALPHABET[i].toInt()] = i
            }
        }

        /**
         * Translates the specified Base64 string into a byte array.
         *
         * @param s the Base64 string (not null)
         * @return the byte array (not null)
         */
        fun decode(s: String): ByteArray {
            val delta = if (s.endsWith("==")) 2 else if (s.endsWith("=")) 1 else 0
            val buffer = ByteArray(s.length * 3 / 4 - delta)
            val mask = 0xFF
            var index = 0
            var i = 0
            while (i < s.length) {
                val c0 = toInt[s[i].toInt()]
                val c1 = toInt[s[i + 1].toInt()]
                buffer[index++] = (c0 shl 2 or (c1 shr 4) and mask).toByte()
                if (index >= buffer.size) {
                    return buffer
                }
                val c2 = toInt[s[i + 2].toInt()]
                buffer[index++] = (c1 shl 4 or (c2 shr 2) and mask).toByte()
                if (index >= buffer.size) {
                    return buffer
                }
                val c3 = toInt[s[i + 3].toInt()]
                buffer[index++] = (c2 shl 6 or c3 and mask).toByte()
                i += 4
            }
            return buffer
        }

    }
}
